nway/utils.py

```diff
@@ -3,7 +3,92 @@
 import PIL.Image
 import os
 import json
-from typing import List, Any, Tuple
+import pandas as pd
+import matplotlib.pyplot as plt
+from pathlib import Path
+from typing import List, Any, Tuple, Optional
+
+
+def summarize_registration_success(pair_outputs):
+    """checks which pairs succeeded in registratio. tries to keep
+    dataframe small for logging by separating column/index names
+    and returning an idmap also
+
+    Parameters
+    ----------
+    pair_outputs: List[dict]
+        at minimum, each entry is {
+                                     'fixed_experiment': <int>,
+                                     'moving_experiment': <int>,
+                                     'transform': {
+                                                    'best_registration': []
+                                                  }
+                                   }
+
+    Returns
+    -------
+    df: pd.DataFrame
+        boolean-valued dataframe, sorted both in rows and columns
+        row/column names replaced with integers
+    id_map: dict
```

Comment:
Just a thought, but I wonder if it might be worthwhile to define a custom dataframe class that inherits from `pd.DataFrame` that contains custom attributes:

https://pandas.pydata.org/pandas-docs/stable/development/extending.html#define-original-properties
or
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.attrs.html

Then you could just return only `df` with the `id_map` tagging along as an attribute of the `df`.

Reply:
i think i'll leave this for now. I see the proposed value, but, I think you'll also shortly be making some DataFrame capabiity for output, and I'll wait to see whether this can then adapt to that.